bottomless/src/replicator.rs

```diff
@@ -693,23 +682,46 @@ impl Replicator {
         tracing::debug!("Rolled back to {}", last_valid_frame);
     }
 
+    // Opens a raw libSQL connection that doesn't checkpoint.
+    // Useful for reading metadata from the database file.
+    fn open_db(&self) -> Result<libsql_sys::Connection<libsql_sys::wal::Sqlite3Wal>> {
+        use libsql_sys::connection::OpenFlags;
+        use libsql_sys::wal::Sqlite3WalManager;
+        let flags = OpenFlags::SQLITE_OPEN_READ_WRITE
+            | OpenFlags::SQLITE_OPEN_CREATE
+            | OpenFlags::SQLITE_OPEN_URI
+            | OpenFlags::SQLITE_OPEN_NO_MUTEX;
+
+        let conn = libsql_sys::Connection::open(
+            &self.db_path,
+            flags,
+            Sqlite3WalManager::new(),
+            u32::MAX, // no checkpointing
+            self.encryption_key.clone(),
+        )?;
+        Ok(conn)
+    }
+
     // Tries to read the local change counter from the given database file
-    async fn read_change_counter(reader: &mut File) -> Result<[u8; 4]> {
-        let mut counter = [0u8; 4];
-        reader.seek(std::io::SeekFrom::Start(24)).await?;
-        reader.read_exact(&mut counter).await?;
-        Ok(counter)
+    fn read_change_counter(&self) -> Result<[u8; 4]> {
+        let conn = self.open_db()?;
+        let change_counter = conn.db_change_counter().map_err(|rc| {
+            anyhow::anyhow!(
+                "Failed to read local change counter from `{}`: {rc}",
+                self.db_path,
+            )
+        })?;
+        tracing::trace!("Local change counter: {change_counter}");
+
+        Ok(change_counter.to_be_bytes())
     }
 
     // Tries to read the local page size from the given database file
-    async fn read_page_size(reader: &mut File) -> Result<usize> {
-        reader.seek(SeekFrom::Start(16)).await?;
-        let page_size = reader.read_u16().await?;
-        if page_size == 1 {
-            Ok(65536)
-        } else {
-            Ok(page_size as usize)
-        }
+    async fn read_page_size(&self) -> Result<usize> {
+        let conn = self.open_db()?;
+        let page_size = conn.query_row("PRAGMA page_size", (), |r| r.get::<usize, usize>(0))?;
```

Comment:
I'm kinda afraid that in bottomless we're picking page_size from database itself, while replication log uses compile time constant.

Reply:
Technically I think it's ok. In libsql-server we assume a 4KiB page and we stick to that, no plans to change the default. Bottomless can be used with or without libsql-server, and if it's used for standalone databases, it makes sense that it also works with other page sizes.